README.md
# webdebugger

[![Build Status](https://travis-ci.org/vscoder/webdebugger.svg?branch=master)](https://travis-ci.org/vscoder/webdebugger)

This webb app show HTTP headers and other useful information

## How to run

### Docker

Just run docker container from Docker Hub

```shell
docker run -p 8080:8080 vscoder/webdebugger:latest
```
And now open webdebugger at http://localhost:8080/

## Configuration

### Environment variables

App behavior variables:
- `APP_DELAY` reply delay, seconds
- `APP_BGCOLOR` set web-page background color. Default `white`
- `APP_NO_CSS` disable loading external CSS or JS resources (usebla inside of isolated environments). Available values: one of `'true', 'yes', '1'` is `True`, every other value is `False`.

Sentry-related variables:
- `SENTRY_DSN` sentry DSN to send traces. Default empty
TODO:

### Routes

Webdebugger process these URI paths

- `/hello` - always return `Hello World! ^_^`
- `/healthz` - always return `OK`
- `/env/<env_var>` - return value of uppercased os environment variable `env_var`
- `/` and `/<path:path>` - are all other URI paths. Return request and os environment variables.

## Development

### Requirements

First, you must have `make` installed, to simplify all the operations.
If you haven't, see [Makefile](Makefile) for corresponding commands.

Then, you must have `poetry` installed. See https://python-poetry.org/docs/

Or just run 
```shell
make install_poetry
```

After that, install dependencies
```shell
make requirements
```
or
```shell
poetry install --local
```

### Run tests

#### pycodestyle

```shell
make pycodestyle
```
or
```shell
poetry run pycodestyle webdebugger/
```

#### pytest

```shell
make pytest
```
or
```shell
poetry run pytest -v
```

#### hadolint (lint Dockerfile)

```shell
make hadolint
```

## Version control

This project use sematic versioning. See [semver](https://semver.org/)

For increasee version, use corresponding Makefile targets
- `version` - get version
- `version-patch` - increase version patch number
- `version-minor` - increase minor version
- `version-major` - increase major version

**NOTE:** All chacges must be committed. All these actions, except `version`, create new commit and add tag with version number.

## Sentry integration

TODO

## Run in docker container

### Run with docker-compose

- `docker-compose up`

App will be available on you'r IP port `8000`

### Variables (`make` args)

Make args are the same as environment variables

### Commands (`make` targets)

Build image
```shell
make build
```

Build image and run tests
```shell
make docker-pytest
```

Run container
```shell
make docker-run
```

Run shell inside container
```shell
make docker-shell
```

## Author

Aleksey Koloskov <vsyscoder@gmail.com>


CHANGELOG.md
## 2021-03-16

- Log request environ to stdout
- Hadolint ignore non strict versions for apk and pip

## 2020-12-13

- Upd sentry release version publication
- Upd openssl-dev to `1.1.1i-r0`

## 2020-11-30

- Add `SENTRY_URL=http://sentry.local:19000/` variable to Makefile
- Add Makefile target `push`
- Add `VERSION.txt` with current version
- Add Makefile targets to manipulate sentry releases
  - sentry-release-new
  - sentry-set-commits
  - sentry-release-finalize

## 2020-11-29

- Add boolean variable `APP_NO_CSS` variable to disable loading external CSS or JS resources (usable in isolated environments)
- Migrate to jinja2 template engine

## 2020-11-22

- Fix docker build (upd mysl-dev vesrsion to `1.1.24-r10`)
- Add sentry integration (use env `SENTRY_DSN`)
- Pass app version as release info to sentry

## 2020-10-04

- Add possibility setting web-page background color with env variable `APP_BGCOLOR`
- Fix Makefile target `hadolint`. Now it works
- Omit variable `APP_PORT` as unusable
- 

## 2020-05-17

- Remove `-${TRAVIS_BUILD_NUMBER}` from docker image tag

## 2020-05-16

- Display webdebugger version in webpage title
  - Add `APP_VERSION` to `Dockerfile`
  - Automatically update `APP_VERSION` in `Dockerfile` by _bump2version_

## 2020-05-15

- Include bootstrap-4.5.0 into html template

## 2020-05-03

- Add route `/env/<env_var>`. Return os environmant variable `env_var` value
- Add route `/healthz`. Always return OK
- Upldate `docker-compose.yml`. Mount `./` as `/app` into container
- Add pycodestyle to `.travis.yml` pipeline

## 2020-04-27

- Add `bump2version` to dev dependencies
- Add `autopep8` to dev dependencies
- Add Makefile target `hadolint`
- Add Makefile target `pycodestyle`
- Improve code style via autopep8
- Add Makefile target `docker-pycodestyle`
- Add Makefile target `docker-version`
- Add Makefile targets for version manipulation
  - `version` - get version
  - `version-patch` - increase version patch number
  - `version-minor` - increase minor version
  - `version-major` - increase major version
- Add `.travis.yml`
- Display version after increase in Makefile targets `version-*`
- Rename script `docker-push.sh` to `docker-publish.sh`
- Don't publish docker image on `maste` branch update
- Add build status badge to `README.md`
- Add `LICENSE.txt`
- Add variables to `.travis.yml`
  - `DOCKER_USERNAME`
  - `DOCKER_PASSWORD`
- Supress debug output in `docker-publish.sh`
- Add Makefile target `docker-publish` to publish docker image via CI
- Set exit with error if any variable value isn't set in `docker-entrypoint.sh`
- Dont do `build` before any of `docker-*` Makefile target
- Add `EXPOSE ${APP_PORT}` instruction to Dockerfile

## 2020-04-26

- Add `docker-compose.yml`

## 2020-04-25

- Remove `request.headers` from output
- Add `os.environ` to output
- Add reply delay (in seconds) with `APP_DELAY` environment variable
- Rename env var `PORT` to `APP_PORT`
- Dockerfile refactoring (spit environment and app configuration env variables)
- Add variable `APP_DELAY`
- Update `Makefile`. Run target build before any `docker-*` target


## 2020-04-24

- Project created


tests/test_webdebugger.py
from webdebugger import __version__


def test_version():
    assert __version__ == '0.12.1'


webdebugger/main.py
import argparse
import logging
import os
from pprint import pformat
from time import sleep

import bottle
import sentry_sdk
from bottle import Bottle, jinja2_view, request, route, run
from sentry_sdk.integrations.bottle import BottleIntegration

FORMAT = '%(message)s'
logging.basicConfig(format=FORMAT)
logger = logging.getLogger(__name__)

bottle.TEMPLATE_PATH.insert(0, "%s/views" % (os.path.dirname(__file__)))


@route('/hello')
def hello():
    return "Hello World! ^_^"


@route('/healthz')
def healthz():
    return "OK"


@route('/exception')
def exception():
    raise Exception("test exception generated")


@route('/env/<env_var>')
def env_var(env_var):
    """
    Return the value of <env_var> os environment variable
    """
    var_value = os.environ.get(env_var.upper())
    return var_value


@route('/')
@route('/<path:path>')
@jinja2_view('info')
def path(path="/"):
    """
    Render main template with lot of request and os information
    """
    # Sleep if defined
    timeout = os.getenv('APP_DELAY')
    if timeout:
        sleep(int(timeout))

    # Collect info
    bottle_env = dict(request.environ)
    os_env = dict(os.environ)

    for var, value in bottle_env.items():
        logger.warning(f'{var} = {value}')
    logger.warning('-'*16)

    return dict(path=path, os_env=os_env, bottle_env=bottle_env)


def main():
    parser = argparse.ArgumentParser(description="Display HTTP headers")
    parser.add_argument("--host", "-H", type=str,
                        default="0.0.0.0", help="listen host")
    parser.add_argument("--port", "-p", type=int,
                        default=8080, help="listen port")
    args = parser.parse_args()

    run(host=args.host, port=args.port, debug=True, reloader=True)


if __name__ == '__main__':
    main()


# Sentry integration args
sentry_dsn = os.getenv('SENTRY_DSN')
release = os.getenv('APP_VERSION')
if sentry_dsn:
    sentry_sdk.init(
        dsn=sentry_dsn,
        release=f"webdebugger@{release}",
        integrations=[BottleIntegration()]
    )

app = bottle.default_app()
